Меняй-удаляй

skudinva opened the pull request:
Заработало редактирование/добавление/удаление/сортировка/фильтрация

Не очень понятно как пользоваться в нашем приложении использовать he.encode. Ведь мы все выбираем из списка.

---
:mortar_board: [Меняй-удаляй](https://up.htmlacademy.ru/ecmascript/23/user/2392413/tasks/15)

skudinva committed:
Исправили сортировку

skudinva committed:
Прикрутил Observable и поправил рендер сортировки

skudinva committed:
Реализовал удаление

skudinva committed:
--consol.log

skudinva committed:
Время в 24часовом формате

skudinva committed:
Не стал глушить линтер

skudinva committed:
Заработала фильтрация

skudinva committed:
Зачистил лишнее

skudinva committed:
рефакторинг

skudinva committed:
Для пустого списка вывожу правильный текст

skudinva committed:
обновляю итоги

skudinva committed:
Доработал добавление

skudinva committed:
++

skudinva committed:
++he

fyvfyv commented on src/presenter/event-presenter.js:
```diff
@@ -89,10 +99,27 @@ export default class EventPresenter {
+      this.#renderTripBoard();
+    } else if (updateType === UpdateType.MAJOR) {
+      this.#clearTripBoard();
+      this.#renderTripBoard();
```
Дублируется код. Кажется, что для major нужно что-то сделать ещё, например, сбросить сортировку

fyvfyv commented on src/presenter/event-presenter.js:
```diff
@@ -89,10 +99,27 @@ export default class EventPresenter {
+  };
+
+  #onEventDataChange = (actionType, updateType, event) => {
+    if (actionType === UserAction.UPDATE_EVENT) {
+      this.#eventsModel.updateEvent(updateType, event);
+    } else if (actionType === UserAction.ADD_EVENT) {
+      this.#eventsModel.addEvent(updateType, event);
+    } else if (actionType === UserAction.DELETE_EVENT) {
+      this.#eventsModel.deleteEvent(updateType, event);
+    }
```
`switch/case`?

skudinva replied:
так терпеть не могу switch, что кушать не могу

fyvfyv replied:
Убедительный аргумент 🤣

fyvfyv commented on src/presenter/event-presenter.js:
```diff
@@ -135,27 +162,27 @@ export default class EventPresenter {
   #renderTripPoints = () => {
-    for (let i = 0; i < this.#events.length; i++) {
-      this.#renderTripPoint(this.#events[i], EditFormMode.EDIT);
+    for (let i = 0; i < this.events.length; i++) {
```
`forEach`

fyvfyv commented on src/view/sort-view.js:
```diff
@@ -1,46 +1,62 @@
+    createSortItemTemplate('offer', 'Offers', currentSortType, true)
+  );
+
+  sortItems.push('</form>');
```
Это можно собрать в массив и запустить цикл

fyvfyv commented on src/view/sort-view.js:
```diff
@@ -1,46 +1,62 @@
+<input id="sort-${type}" class="trip-sort__input  visually-hidden" type="radio" name="trip-sort" value="sort-${type}" ${additionProps}>
+<label class="trip-sort__btn" for="sort-${type}">${label}</label>
+</div>`;
+};
```
Форматирование поехало

skudinva committed:
bug

fyvfyv commented on src/presenter/event-point-presenter.js:
```diff
@@ -108,14 +108,32 @@ export default class EventPointPresenter {
       offersList: this.#offersList,
-      onSubmit: (updateEvent) => {
-        this.#handleDataChange(updateEvent);
+      onFormSubmit: (updateEvent) => {
+        if (isNewEvent(updateEvent)) {
+          updateEvent.id = guid();
+          this.#handleDataChange(
+            UserAction.ADD_EVENT,
+            UpdateType.MAJOR,
+            updateEvent
+          );
+        } else {
+          this.#handleDataChange(
+            UserAction.UPDATE_EVENT,
+            UpdateType.MINOR,
+            updateEvent
+          );
+        }
         this.#handleStateChange(this, EventStateAction.SUBMIT_EDIT_FORM);
       },
```
Я бы всё-таки подумал о разделении формы добавления и редактирования.

skudinva committed:
При изменении фильтров сбрасываю сортировку

skudinva committed:
if => switch/case

skudinva committed:
for => forEach

skudinva committed:
Собрал в массив и запустил цикл

skudinva committed:
форматирование

skudinva committed:
Устранил замечания ПР. Одно оставил на подумать.

fyvfyv commented:
По поводу `he.encode()`. Он у тебя падал, потому что ты подавал ему на вход `cityName: undefined`, а там внутри нет проверки на то, что передали не строку.
Если в `cityName` пустого эвента будет не undefined, а пустая строка, то всё сработает.

По поводу использования на выводе или вводе: можно просто обернуть в него `value` у инпутов, чтобы на уровне логики во вьюхах/сервисах вообще не думать об этом
`<input ... value="${he.encode(cityName)}"`

skudinva committed:
Поправил баг с фильтрами.
При изменении фильтра сбрасываю сортировку.
При добавлении сбрасываю сортировку и фильтр.

skudinva committed:
dateFrom < dateTo

skudinva committed:
++he

skudinva committed:
В basePrice можно вводить только цифры

skudinva committed:
Если пустую карточку сохраняешь, то валится

skudinva committed:
he fix
